File: code/0307/307.区域和检索-数组可修改.cpp
/*
 * @lc app=leetcode.cn id=307 lang=cpp
 * @lcpr version=21302
 *
 * [307] 区域和检索 - 数组可修改
 */

// @lc code=start
class NumArray {
public:
    int n;
    vector<int> tr, nums;

    int lowbit(int x) {
        return x & -x;
    }

    int query(int x) {
        int res = 0;
        for (int i = x; i; i -= lowbit(i))
            res += tr[i];
        return res;
    }

    void add(int x, int v) {
        for (int i = x; i <= n; i += lowbit(i))
            tr[i] += v;
    }

    NumArray(vector<int>& _nums) {
        nums = _nums;
        n = nums.size();
        tr.resize(n + 1);
        for (int i = 0; i < n; i++)
            add(i + 1, nums[i]);
    }
    
    void update(int index, int val) {
        add(index + 1, val - nums[index]);
        nums[index] = val;
    }
    
    int sumRange(int left, int right) {
        return query(right + 1) - query(left);
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * obj->update(index,val);
 * int param_2 = obj->sumRange(left,right);
 */
// @lc code=end



/*
// @lcpr case=start
// 
// @lcpr case=end

 */

File: code/0307/307.区域和检索-数组可修改.py
#
# @lc app=leetcode.cn id=307 lang=python3
# @lcpr version=21302
#
# [307] 区域和检索 - 数组可修改
#

# @lc code=start
class NumArray:
    def __init__(self, nums: List[int]):
        self.num = nums
        self.n = len(nums)
        self.tr = [0] * (self.n + 1)
        for i, x in enumerate(nums, 1):
            self.add(i, x)

    def update(self, index: int, val: int) -> None:
        self.add(index + 1, val - self.num[index])
        self.num[index] = val

    def sumRange(self, left: int, right: int) -> int:
        return self.query(right + 1) - self.query(left)
    
    def lowbit(self, x: int) -> int:
        return x & -x

    def add(self, x: int, v: int) -> None:
        i = x
        while i <= self.n:
            self.tr[i] += v
            i += self.lowbit(i)
    
    def query(self, x: int) -> int:
        res, i = 0, x
        while i:
            res += self.tr[i]
            i -= self.lowbit(i)
        return res



# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)
# @lc code=end



#
# @lcpr case=start
# 
# @lcpr case=end

#


